Speech recognition example - sequence to sequence using transformer

apoorvnandan opened the pull request:
Some decisions I made to keep this example short and easy to train:

- Model - Transformer (both encoder and decoder):
I picked a sequence to sequence approach instead of a CTC pipeline because we already have a good example showing the CTC setup. (OCR Captcha Recognition) and a CTC based example will be very similar to it.

- Data - Free spoken digit dataset:
A standard (and more realistic) dataset like LibriSpeech requires few hours for training (even with a V100) to reach a WER < 10%. Free spoken digit contains small segments of audio (<= 1 second) and a single word as the transcript (treated as a sequence of characters). This data is very easy to train on but demonstrates a complete sequence to sequence pipeline.
 
- To make it slightly more difficult I prepared another dataset from this, by combining 2 random audio files and their transcripts, to create longer audio segments and character sequences. This requires more iterations to achieve good performance.

- I was not sure if the extra lines of code to create the modified dataset would be useful for this demo, so I have committed the script with the original one-word dataset.

- I am aware that a real-world speech recognition system involves a lot more bells and whistles (especially with a transformer model), but this is just a minimal example.

fchollet commented:
Thank you for the PR! This will make an awesome example. We definitely needed a Transformer-based sequence-to-sequence model.

> A standard (and more realistic) dataset like LibriSpeech requires few hours for training (even with a V100) to reach a WER < 10%. Free spoken digit contains small segments of audio (<= 1 second) and a single word as the transcript (treated as a sequence of characters). This data is very easy to train on but demonstrates a complete sequence to sequence pipeline.

Training time is not a problem: you can simply write down in a comment the config that reaches convergence, but keep the training in the example to a single epoch. You can then report the evaluation results as a comment. Running the example should be fast, but the example doesn't have to converge with its default config.

For instance, in [this GAN example](https://keras.io/examples/generative/dcgan_overriding_train_step/), we just train for 1 epoch, but recommend training for ~100 epochs. We display the results we get after 30 epochs.

> Our data consists of audio segments with a person speaking out one of the ten digits
> (0-9). We convert the digits as a sequence of characters (e.g. 9 -> n,i,n,e) to form
> a small dataset

I think it would be a lot more interesting (and more standards) to have the output tokens be words, not characters. Consider using LibriSpeech (or similar) even if convergence takes too long. The task above is basically no different from audio sample classification.

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,391 @@
+        self.input_type = input_type
+        if input_type == "tokens":
+            self.emb = tf.keras.layers.Embedding(nvocab, nhid)
+        elif input_type == "feats":
```
In general, avoid abbreviations, e.g. you can use "features"

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,391 @@
+        if input_type == "tokens":
+            self.emb = tf.keras.layers.Embedding(nvocab, nhid)
+        elif input_type == "feats":
+            self.emb = tf.keras.layers.Dense(nhid)
```
You might want to handle the 3rd case with an error or otherwise have some form of value validation for "input_type"

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,391 @@
+
+class TransformerInput(layers.Layer):
+    def __init__(
+        self, input_type="tokens", nvocab=1000, nhid=64, nff=128, maxlen=100,
```
Standard number prefix is `num_`

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,391 @@
+    def call(self, x):
+        if self.input_type == "tokens":
+            maxlen = tf.shape(x)[-1]
+        elif self.input_type == "feats":
```
Once the value is validated you can just use if/else

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,391 @@
+            y = getattr(self, f"dec_layer_{i}")(enc_out, y)
+        return y
+
+    def call(self, inputs):
```
Consider having 2 args in the signature

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,391 @@
+        trainable_vars = self.trainable_variables
+        gradients = tape.gradient(loss, trainable_vars)
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+        return {"loss": loss}
```
For loss tracking, use a `Mean` metric. Otherwise you're displaying the per-batch loss (high variance) instead of the running average of the loss over each epoch (more stable). Also create a `metrics` `@property` listing the tracker, so the model is able to reset the average after every epoch. See e.g. the GAN example.

apoorvnandan commented:
Sorry for the delay. I ran a bunch of experiments and made the following changes:
- Words vs Character as targets
> I think it would be a lot more interesting (and more standards) to have the output tokens be words, not characters. 

I was both curious and skeptical about this since I have never seen any ASR paper or production system use words as the targets. (Always characters, phonemes, or BPE). 

But I tried it out, and even though it reduces the output sequence length, it takes significantly longer for the network to even overfit on a small subset of the data, let alone converge on the full dataset. Plus the exposure bias tends to be extreme. The decoder relies heavily on the past ground truth tokens during training to bring the loss down. So I picked characters.

- Changed dataset to [LJSpeech](https://keithito.com/LJ-Speech-Dataset/). It has up to 10-second long audio snippets and takes up ~3.6 GB on the disk.
- Added `Conv1D` layers before the encoder stack. These serve dual purposes: (i) reducing sequence length and (ii) extracting useful features from the spectrograms. Overall, it is challenging to get decent performance without them.
- Added a learning rate schedule (significantly speeds up the training) and label smoothing.

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+"""
+
+
+class TransformerDecoderLayer(layers.Layer):
```
Likewise, `TransformerDecoder`

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+        )
+        self.enc_att = layers.MultiHeadAttention(num_heads=num_heads, key_dim=embed_dim)
+        self.self_drop = layers.Dropout(0.5)
+        self.enc_drop = layers.Dropout(0.0)
```
A dropout layer with no dropout does nothing, you can remove it, or otherwise set the rate to a proper value

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+"""
+
+
+class TransformerInput(layers.Layer):
```
The use of `Input` can be confusing due to the collision with `keras.Input`. I would recommend calling this `PositionalEmbedding` or something similar, as well as separating the code for the 2 input types, since in practice they have no overlap. This class will be more readable if it only covers the "tokens" mode.

The "features" mode probably doesn't require its own class, you could just inline the conv1d layers in the `Transformer` class.

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+)
+
+
+def masked_loss(real, pred):
```
You don't really need this, instead you can rely on the `sample_weight` argument of `CategoricalCrossentropy` when you use it in `train_step`, e.g.

```
one_hot = tf.one_hot(y_true, depth=self. num_classes)
mask = tf.math.logical_not(tf.math.equal(y_true, 0))
loss = loss_fn(y_true, y_pred, sample_weight=mask)
```

(and enable reduction in the loss)

Also avoid hardcoding parameters like "34"

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+            num_hid=num_hid,
+            maxlen=trg_maxlen,
+        )
+        for i in range(num_layers_enc):
```
Use a Sequential (with a list comprehension) instead of setattr/getattr, it will be much simpler.

```
self.encoder = keras.Sequential(
   [PositionalEmbedding(...)] + [TransformerEncoder(...) for _ in range(num_encoder_layers)]
)
```
then you don't even the encoder/decoder methods, just do e.g. `self.encoder(inputs)`

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+        dec_trg = trg[:, 1:]
+        with tf.GradientTape() as tape:
+            preds = self([src, dec_inp])
+            loss = masked_loss(dec_trg, preds)
```
In general the inner code of a model should not rely on variables that aren't attributes of the model (for the model to be portable it needs to be self-contained). Here the loss should be configured in compile, so that you do e.g. `self.compiled_loss(y_true, y_pred, sample_weight=mask)`

or you could pass the loss in the constructor and set it as an attr.

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+                TransformerDecoderLayer(num_hid, num_head, num_ff),
+            )
+
+        self.final = layers.Dense(num_classes)
```
classifier

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+max_target_len = 200  # all transcripts in out data are < 200 characters
+data = get_data(wavs, id_to_text, max_target_len)
+vectorize_layer = VectorizeChar(max_target_len)
+len(vectorize_layer.get_vocabulary())
```
Always use `print()` (and a formatted message) to display something

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+
+max_target_len = 200  # all transcripts in out data are < 200 characters
+data = get_data(wavs, id_to_text, max_target_len)
+vectorize_layer = VectorizeChar(max_target_len)
```
It's not a layer, just `vectorizer`

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+            print()
+
+
+def scheduler(epoch, lr):
```
It would be simpler to use one of the built-in learning rate schedules, like this one: https://keras.io/api/optimizers/learning_rate_schedules/inverse_time_decay/

apoorvnandan replied:
I could not find a built-in schedule for warmup. https://keras.io/api/optimizers/learning_rate_schedules/
A simple decay might lead to a performance drop or instability. I've observed warmups to be particularly important with transformers.

fchollet replied:
You could structure your own schedule as a learning rate schedule -- just a callable class where the `__call__` takes a `step` arg.

See https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/optimizer_v2/learning_rate_schedule.py

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+
+
+for i in val_ds.take(1):
+    batch = i  # Use the first batch of validation set to display outputs
```
The purpose of these 2 lines is unclear, please clarify

apoorvnandan replied:
These are just to fetch a single batch from the validation set for the display callback.
It can be replaced by `batch = next(iter(val_ds))[1]`, if that feels better.

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+)
+model.compile(optimizer="adam")
+
+history = model.fit(ds, callbacks=[display_cb, schedule_cb], epochs=1)
```
We should be monitoring a validation dataset during training (the text display callback is not sufficient for monitoring, we should be also looking at the validation loss to check overfitting)

apoorvnandan replied:
Okay, overriding `test_step` for this.

fchollet commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+    # normalisation
+    means = tf.math.reduce_mean(x, 1, keepdims=True)
+    stddevs = tf.math.reduce_std(x, 1, keepdims=True)
+    x = tf.divide(tf.subtract(x, means), stddevs)
```
why not `/` and `-`?

fchollet reviewed (commented):
Thanks for the update! Things are looking good. I added a larger batch of comments -- I think this should be everything.

fchollet commented:
Hi Apoorv, are you still working on this? Let me know if there's any help I can provide. It's an awesome example, would love to see it included. Cheers!

apoorvnandan commented:
Hi. Sorry, these last few weeks were extremely hectic for me. 
I'll make the changes this weekend.

fchollet commented:
Take your time -- it's ready when it's ready. I just wanted to make sure you weren't blocked by anything.

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+"""
+## Introduction
+
+Automatic speech recognition (ASR) consists of transcribing audio segments into text.
```
Suggestion:

```suggestion
Automatic speech recognition (ASR) tasks involve turning speech audio signals into text.
```

- Speech audio is not the same as music audio (the latter is MIR-related)

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+## Introduction
+
+Automatic speech recognition (ASR) consists of transcribing audio segments into text.
+A popular way to handle this task is to treat it as a sequence-to-sequence
+problem. The audio can be represented as a sequence of feature vectors,
+and the text can be represented as a sequence of characters, words, or subword tokens.
+
```
Suggestion:

```suggestion
ASK tasks can be treat sequence-to-sequence problems, where the
audio can be represented as a sequence of feature vectors
and the text—as a sequence of characters, words, or subword tokens.

```

- In ASR you work on structured sequential data, so ASR systems you build map sequences to sequences. Hence, I'd suggest you say that this is a sequence-to-sequence (referring to the approach developed in 2014) approach "and the audio ..."

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+problem. The audio can be represented as a sequence of feature vectors,
+and the text can be represented as a sequence of characters, words, or subword tokens.
+
+For this demonstration, we will use LJSpeech dataset from the
+[LibriVox](https://librivox.org/) project. It consists of short
+audio clips of a single speaker reading passages from 7 non-fiction books.
+Our model will be similar to the original Transformer (both encoder and decoder)
+as proposed in the paper, "Attention is All You Need".
+
+
+**References:**
+- [Attention is All You Need](https://papers.nips.cc/paper/2017/file/3f5ee243547dee91fbd053c1c4a845aa-Paper.pdf)
+- [Very Deep Self-Attention Networks for End-to-End Speech Recognition](https://arxiv.org/pdf/1904.13377.pdf)
+- [Speech Transformers](https://ieeexplore.ieee.org/document/8462506)
+- [LJSpeech Dataset](https://keithito.com/LJ-Speech-Dataset/)
```
Suggestion: maybe this can help save some space:

```suggestion
For this demonstration, we will use [LJSpeech](https://keithito.com/LJ-Speech-Dataset/) dataset from the
[LibriVox](https://librivox.org/) project. It consists of short
audio clips of a single speaker reading passages from seven non-fiction books.
The model you will use is similar to the original encoder-decoder Transformer model,
as proposed in Attention is All You Need ([Vaswani et al., 2017](https://papers.nips.cc/paper/2017/file/3f5ee243547dee91fbd053c1c4a845aa-Paper.pdf)). You can also refer to Very Deep Self-Attention Networks for End-to-End Speech Recognition ([Pham et al, 2019](https://arxiv.org/pdf/1904.13377.pdf)) and Speech-Transformer: A No-Recurrence Sequence-to-Sequence Model for Speech Recognition ([Dong et al., 2018](https://ieeexplore.ieee.org/document/8462506)).
```

(Note the IEEE link is "paywalled".)

8bitmp3 reviewed (commented):
My 2 cents:

```suggestion
Description: Training a sequence-to-sequence Transformer for automatic speech recognition.
```

- ASR is defined in the first paragraph after this.
- Though the original seq2seq paper didn't use dashes ("-"), most other literature I've seen uses "seq..-to-seq..".
- Also, see @fchollet 's example https://keras.io/examples/vision/image_classification_from_scratch/ where the description starts with "Training...." (i.e. "doing something" instead of "do something")

---


Suggestion:

```suggestion
Automatic speech recognition (ASR) tasks involve turning speech audio signals into text.
```

- Speech is the key word - e.g. there's speech audio and music audio (the latter is Music Information Retrieval-related)

---

Suggestion:

```suggestion
ASK tasks can be treat sequence-to-sequence problems, where the
audio can be represented as a sequence of feature vectors
and the text—as a sequence of characters, words, or subword tokens.

```

- In ASR you work on structured sequential data, so ASR systems you build map sequences to sequences. Hence, I'd suggest you say that this is a sequence-to-sequence (referring to the approach developed in 2014) approach "and the audio ..."

---

Suggestion: maybe this can help save some space:

```suggestion
For this demonstration, we will use [LJSpeech](https://keithito.com/LJ-Speech-Dataset/) dataset from the
[LibriVox](https://librivox.org/) project. It consists of short
audio clips of a single speaker reading passages from seven non-fiction books.
The model you will use is similar to the original encoder-decoder Transformer model,
as proposed in Attention is All You Need ([Vaswani et al., 2017](https://papers.nips.cc/paper/2017/file/3f5ee243547dee91fbd053c1c4a845aa-Paper.pdf)). You can also refer to Very Deep Self-Attention Networks for End-to-End Speech Recognition ([Pham et al, 2019](https://arxiv.org/pdf/1904.13377.pdf)) and Speech-Transformer: A No-Recurrence Sequence-to-Sequence Model for Speech Recognition ([Dong et al., 2018](https://ieeexplore.ieee.org/document/8462506)).
```

(Note the IEEE link is "paywalled". I couldn't find an alternative.)
---

8bitmp3 commented on examples/audio/transformer_asr.py:
````diff
@@ -0,0 +1,523 @@
+history = model.fit(ds, callbacks=[display_cb, schedule_cb], epochs=1)
+
+"""
+In practice, train for ~100 epochs. 
+
+Some of the predicted text around epoch 35.
+```
+target:     <as they sat in the car, frazier asked oswald where his lunch was>
+prediction: <as they sat in the car frazier his lunch ware mis lunch was>
+
+target:     <under the entry for may one, nineteen sixty,>
+prediction: <under the introus for may monee, nin the sixty,>
+```
````
```suggestion
In practice, you should train for around 100 epochs or more. 

Some of the predicted text at or around epoch 35 may look as follows:

```
target:     <as they sat in the car, frazier asked oswald where his lunch was>
prediction: <as they sat in the car frazier his lunch ware mis lunch was>

target:     <under the entry for may one, nineteen sixty,>
prediction: <under the introus for may monee, nin the sixty,>
```
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+for i in val_ds.take(1):
+    batch = i  # Use the first batch of validation set to display outputs
+
+# vocabulary to convert predicted indices to characters
```
```suggestion
# The vocabulary to convert predicted indices into characters
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+            prediction = ""
+            over = False
+            for idx in preds[i, :]:
+                if over:  # Add padding token once end token has beeen predicted
```
```suggestion
                if over:  # Add a padding token once the end token has been predicted
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+    def __init__(
+        self, batch, idx_to_token, trg_start_token_idx=27, trg_end_token_idx=28
+    ):
+        """ Displays a batch of outputs after every epoch 
+        Arguments:
+        - batch: test batch containing the keys "src" and "trg"
+        - idx_to_token: a List containing the vocabulary tokens corresponding to their indices
+        - trg_start_token_idx: start token index in the target vocabulary
+        - trg_end_token_idx: end token index in the target vocabulary
+        """
```
Style guide https://google.github.io/styleguide/pyguide.html


```suggestion
        """ Displays a batch of outputs after every epoch.
        
        Args:
          batch: A test batch containing the keys "src" and "trg".
          idx_to_token: A list containing the vocabulary tokens corresponding to their indices.
          trg_start_token_idx: A start token index in the target vocabulary.
          trg_end_token_idx: An end token index in the target vocabulary.
        """
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+
+
+"""
+## Callbacks to display predictions and to change learning rate
```
```suggestion
## Callbacks to display predictions and to change the learning rate
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+
+
+"""
+Check contents of one batch
```
```suggestion
Check the contents of one batch
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+
+
+"""
+## Preprocess dataset
```
```suggestion
## Preprocess the dataset
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+
+
+"""
+## Download dataset
```
```suggestion
## Download the dataset
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+        return {"loss": self.loss_metric.result()}
+
+    def generate(self, src, trg_start_token_idx):
+        """ inference for one batch of inputs using greedy decoding """
```
```suggestion
        """Performs inference over one batch of inputs using greedy decoding."""
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+        return [self.loss_metric]
+
+    def train_step(self, batch):
+        """ Process one batch inside model.fit() """
```
```suggestion
        """Processes one batch inside model.fit()."""
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+
+
+def masked_loss(real, pred):
+    """ assuming pad token index = 0 """
```
Suggestion:

- Add a doctring explaining what the function does

```suggestion
    """Computes the masked loss assuming the pad token index is 0."""
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+
+
+"""
+## Complete Transformer Model
```
```suggestion
## Complete the Transformer model
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+"""
+## Complete Transformer Model
+
+Our model takes audio spectrograms as input, and predicts a sequence of characters.
+During training, we give the decoder the target character sequence shifted to the left
+as input. During inference, the decoder uses its own past predictions to predict the 
+next token.
```
```suggestion
Our model takes audio spectrograms as inputs and predicts a sequence of characters.
During training, we give the decoder the target character sequence shifted to the left
as input. During inference, the decoder uses its own past predictions to predict the 
next token.
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+        )
+
+    def causal_attention_mask(self, batch_size, n_dest, n_src, dtype):
+        """Mask the upper half of the dot product matrix in self attention.
+
+        This prevents flow of information from future tokens to current token.
+        1's in the lower triangle, counting from the lower right corner.
+        """
```
https://google.github.io/styleguide/pyguide.html

```suggestion
        """Masks the upper half of the dot product matrix in self attention.

        This prevents flow of information from future tokens to current token.
        1's in the lower triangle, counting from the lower right corner.
        """
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,523 @@
+
+
+"""
+## Transformer Input Layer
```
```suggestion
## Define the Transformer's input layer
```

apoorvnandan commented:
I've made the changes. @fchollet Kindly review once.

Thanks, @8bitmp3 for the suggestions!

fchollet reviewed (approved):
LGTM, thank you for the great contribution! Code patterns look good. Applied some copyedits, pull them first. You can add the autogenerated files. 👍

apoorvnandan commented:
I'm getting this error `ValueError: Found 365 lines of code, but expected fewer than 300`. I guess I can remove the whole display outputs callback and the `generate` function (greedy decoding on transformer outputs)? Any suggestions?

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,555 @@
+audio can be represented as a sequence of feature vectors
+and the text as a sequence of characters, words, or subword tokens.
+
+For this demonstration, we will use the LJSpeech dataset from the
+[LibriVox](https://librivox.org/) project. It consists of short
+audio clips of a single speaker reading passages from 7 non-fiction books.
+Our model will be similar to the original Transformer (both encoder and decoder)
+as proposed in the paper, "Attention is All You Need".
+
+
+**References:**
+
+- [Attention is All You Need](https://papers.nips.cc/paper/2017/file/3f5ee243547dee91fbd053c1c4a845aa-Paper.pdf)
+- [Very Deep Self-Attention Networks for End-to-End Speech Recognition](https://arxiv.org/pdf/1904.13377.pdf)
+- [Speech Transformers](https://ieeexplore.ieee.org/document/8462506)
+- [LJSpeech Dataset](https://keithito.com/LJ-Speech-Dataset/)
```
To help you with shrinking the example down to 300 lines, maybe you can:

- Use one line per paragraph, since it's all Markdown here
- Cite the references directly in the text

```suggestion
For this demonstration, we will use the [LJSpeech](https://keithito.com/LJ-Speech-Dataset/) dataset from the [LibriVox](https://librivox.org/) project, which consists of short audio clips of a single speaker reading passages from 7 non-fiction books. Our model will be similar to the [original Transformer](https://papers.nips.cc/paper/2017/file/3f5ee243547dee91fbd053c1c4a845aa-Paper.pdf) (both encoder and decoder) model (Vaswani et al., 2017) (also see [Pham et al., 2019]((https://arxiv.org/pdf/1904.13377.pdf)) and [Speech Transformers](https://ieeexplore.ieee.org/document/8462506) (Dong et al., 2019)
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,555 @@
+"""
+## Introduction
+
+Automatic speech recognition (ASR) consists of transcribing audio speech segments into text.
+ASR can be treated as a sequence-to-sequence problem, where the
+audio can be represented as a sequence of feature vectors
+and the text as a sequence of characters, words, or subword tokens.
```
To help you with shrinking the example down to 300 lines, maybe you can:

- Use one line per paragraph, since it's all Markdown here

```suggestion
Automatic speech recognition (ASR) consists of transcribing audio speech segments into text. ASR can be treated as a sequence-to-sequence problem, where the audio can be represented as a sequence of feature vectors and the text as a sequence of characters, words, or subword tokens.
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,555 @@
+"""
+## Define the Transformer Input Layer
+
+When processing past target tokens for the decoder, we compute the sum of 
+position embeddings and token embeddings.
+
+When processing audio features, we apply convolutional layers to downsample
+them (via convolution stides) and process local relationships.
```
Found a typo: `stides` -> `strides`

---

To help you with shrinking the example down to 300 lines, maybe you can:

- Use one line per paragraph, since it's all Markdown here

```suggestion
When processing past target tokens for the decoder, we compute the sum of position embeddings and token embeddings. When processing audio features, we apply convolutional layers to downsample them (via convolution strides) and process local relationships.
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,555 @@
+
+
+"""
+## Transformer Decoder Layer
```
```suggestion
## Define the Transformer decoder layer
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,555 @@
+
+
+"""
+## Transformer Encoder Layer
```
```suggestion
## Define the Transformer Encoder layer
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,555 @@
+"""
+## Complete the Transformer model
+
+Our model takes audio spectrograms as inputs and predicts a sequence of characters.
+During training, we give the decoder the target character sequence shifted to the left
+as input. During inference, the decoder uses its own past predictions to predict the 
+next token.
```
To help you with shrinking the example down to 300 lines, maybe you can:

Use one line per paragraph, since it's all Markdown here

```suggestion
Our model takes audio spectrograms as inputs and predicts a sequence of characters. During training, the decoder receives the target character sequence shifted to the left as input. During inference, the decoder uses its own past predictions to predict the next token.
```

8bitmp3 commented on examples/audio/transformer_asr.py:
```diff
@@ -0,0 +1,555 @@
+"""
+## Download the dataset
+
+Note: This requires ~3.6 GB of disk space and
+takes ~5 minutes for the extraction of files.
```
To help you with shrinking the example down to 300 lines, maybe you can:

- Use one line per paragraph, since it's all Markdown here

```suggestion
Note: This requires ~3.6 GB of disk space and takes ~5 minutes for the extraction of files.
```

8bitmp3 reviewed (commented):
To help you with shrinking the example down to 300 lines, maybe you can:

- Use one line per paragraph, since it's all Markdown here
- Cite the references directly in the text

I also found a small type (`stride`)

LMKWYT

apoorvnandan commented:
Hi @8bitmp3 There are two problems with what you have suggested.
1. I believe the limit is on the number of lines of code. Not total lines including the descriptions and the texts. Otherwise, the file is ~550 lines, not 365 as shown in the error.
2. The reason I broke those sentences up into multiple short lines of text is that the autogen script is not good at doing so by itself for longer lines.
Quoting from the README
>NOTE THAT THE CONVERSION SCRIPT MAY MAKE MISTAKES IN ITS ATTEMPTS TO SHORTEN LINES. MAKE SURE TO PROOFREAD THE GENERATED .py IN FULL. Or alternatively, make sure to keep your lines reasonably-sized (<90 char) to start with, so that the script won't have to shorten them.

fchollet commented:
> I'm getting this error ValueError: Found 365 lines of code, but expected fewer than 300. I guess I can remove the whole display outputs callback and the generate function (greedy decoding on transformer outputs)? Any suggestions?

Please try to remove any line of code that may not be strictly necessary. If, having done this, the example is still over the limit, we can override the limit. Less code is better, but we don't want to remove anything essential.

> Use one line per paragraph, since it's all Markdown here

Text lines and comment lines do not count (more of them is usually better). Keeping lines under ~90 chars helps greatly with PR reviews. In fact, no amount of superficial reformatting can help bring down the line count, since the code style is controlled by Black. The only way is actually reduce the amount of code.

8bitmp3 commented:
Cool @fchollet @apoorvnandan No worries

apoorvnandan commented:
Removed some unnecessary lines. The file still contains ~350 lines of code, which cannot be brought down to 300 without removing entire sections, e.g. everything related to decoding and displaying output text. 
I generated the `md` and `ipynb` files by changing the limit in `tutobooks.py`. Please check if this works.

fchollet commented:
Excellent, I'll take it from here. Thank you for the great contribution! 👍
